web/packages/teleterm/src/services/tshdEvents/index.ts

```diff
@@ -118,100 +122,104 @@ async function createServer(
 
-// createService creates a service that can be added to tshd events server. It also
-// returns a function which lets UI code subscribe to events which are emitted when a client calls
-// this service.
-//
-// Why do we need to use an event emitter? The gRPC server is created in the preload script but we
-// need the UI to react to the events. We cannot create the service in the UI code because this
-// would mean that the service would need to cross the contextBridge. This is simply impossible
-// because the service is fed custom gRPC classes which can't be passed through the contextBridge.
-//
-// Instead, we create an event emitter and expose subscribeToEvent through the contextBridge.
-// subscribeToEvent lets UI code register a callback for a specific event. That callback receives
-// a simple JS object which can freely pass the contextBridge.
-//
-// # Async behavior
-//
-// The callback can return a promise. The service will not return a response until all callbacks
-// resolve. This lets us model behavior where tshd calls the Electron app and then blocks until it
-// receives a response, in case the Electron app needs to do some work before we want to unblock
-// tshd.
-//
-// If any of the callbacks return an error, the service will return that error immediately, without
-// waiting for other listeners.
+/**
+ * createService creates a service for the tshd events server. It sets up the preload part of the
+ * service and expects the UI to call setupTshdEventContextBridgeService to add the browser part of
+ * the service which interacts with the UI.
+ *
+ * See the JSDoc for TshdEventContextBridgeService for more details.
+ */
 function createService(logger: Logger): {
   service: apiService.ITshdEventsServiceServer;
-  subscribeToTshdEvent: SubscribeToTshdEvent;
+  setupTshdEventContextBridgeService: (
+    listener: TshdEventContextBridgeService
+  ) => void;
 } {
-  const emitter = new Emittery();
+  let contextBridgeService: TshdEventContextBridgeService;
 
-  const subscribeToTshdEvent: SubscribeToTshdEvent = (eventName, listener) => {
-    emitter.on(eventName, listener);
+  const setupTshdEventContextBridgeService = (
+    newListener: TshdEventContextBridgeService
+  ) => {
+    contextBridgeService = newListener;
   };
 
-  const service: apiService.ITshdEventsServiceServer = {
-    relogin: (call, callback) => {
-      const request = call.request.toObject();
-
-      logger.info('Emitting relogin', request);
+  /**
+   * processEvent is a helper function encapsulating common operations done before and after sending
+   * the event to the browser through the context bridge.
+   *
+   * The last argument is a function which maps the response object (received from the browser
+   * through the context bridge) to a class instance (as expected by grpc-js).
+   */
+  function processEvent<
+    RpcName extends keyof apiService.ITshdEventsServiceServer,
+    Request extends protobuf.Message,
+    Response extends protobuf.Message
```

Comment:
`Request` and `Response` could in theory come from `ITshdEventsServiceServer` as well. But it'd make the types just as complex as in `TshdEventContextBridgeService`, which I wanted to avoid, since it's not necessary here. As shown by the code, `processEvent` can type check while working on "abstract" `protobuf.Message`.

Reply:
I played with that trying to make the types more strict, but it was becoming so complex :/

I agree with using just `protobuf.Message`, it's fine here.